### ISSUE TITLE
Clarify the purpose of the stylesheet and the tools

### ISSUE THREAD

**Issue by ApiSecRay** [restful-api-description-language/RADL#3]
_From @ApiSecRay on March 31, 2015 8:15_

Which should do validation?

_Copied from original issue: restful-api-description-language/radl-core#6_

**Comment by jonathanrobie** [comment]
"The tools" is too vague a term in this issue. I would like to restrict this particular issue to the relationship between the stylesheet and the Java program that invokes it. As I see it, the tool that generates document could simply do schema validation, then run the stylesheet, with very little logic in the Java code.

- RELAX-NG validation should be done against the schema.
- The stylesheet should not silently produce wrong output due to referential integrity issues without issuing warnings - users should not have to read through the generated documentation to see if references were wrong in their RADL file.  It should use <xsl:message/> for these conditions, sending warnings to the command line, as the W3C spec stylesheets and DocBook stylesheets do.
- The stylesheet should be usable on its own, not only when invoked via Java. Consistency checks in the stylesheet are useful for that purpose. This is especially important when working on changes to the stylesheet, schema, and instances at the same time.
- For pure XML processing, an XML language like XSLT or XQuery has advantages over Java, and a lot of the required logic is already in the stylesheet.

**Comment by ApiSecRay** [comment]
I prefer the Unix philosophy of lots of little tools that each do one thing and do it well. Cramming everything into a single stylesheet makes that needlessly complex and violates the [Single Responsibility Principle](http://www.objectmentor.com/resources/articles/srp.pdf).

We have two responsibilities here:

1. Generating client documentation from RADL
1. Validating RADL

I propose that we keep those responsibilities separate. I agree that we shouldn't produce documentation based on an invalid service description. But we also shouldn't generate Java code based on an invalid service description. That doesn't mean the code generation tool should do validation as well; it means we should extract the validation functionality into a different "tool" and execute the "tools" in order.

Note that I'm using the word "tool" loosely here. For each "tool", we should decide what the best technology is to implement it. For transforming XML into HTML, that clearly is XSLT. For validation, we can have a discussion about the pros and cons of XSLT versus Java. Whatever we decide there, we should have two "tools", so either two separate stylesheets, or a stylesheet and a Java program.

**Comment by jonathanrobie** [comment]
One possibility is a validation library written in XSLT that could be used by both a validation tool and the stylesheet.

When I'm changing the schema, examples, and stylesheets at the same time, I don't want to have to step out of my XML environment to see any errors, I want my stylesheet to warn me.  If it does so by importing a validation library and using it, that's fine.

**Comment by ApiSecRay** [comment]
Yes, such a hybrid approach that could work as well. But that's a matter of **how** we implement it. I'd like to get agreement on **what** we do first: one stylesheet that does both transformation to HTML and validation, or two separate "tools"?

**Comment by jonathanrobie** [comment]
I don't want the stylesheet to silently produce wrong output unless some other program is run to make sure the input is OK. The testing that is required by the stylesheet is also useful elsewhere. So putting it in a library makes sense so it can be reused in a separate tool, or imported into another stylesheet.

I think this entire issue is about *how* we implement it, using separate programs or library modules or whatever.  We know *what* we want to do, and we know that we want to be able to reuse the logic.

**Comment by jonathanrobie** [comment]
We discussed this today.  We agreed on the following:

- We will separate validation from the stylesheet per se
- Both validation and the stylesheet will be written in XSLT for the sake of platform independence (e.g. Node.js)
- Errors will be reported in a machine-readable format, probably XML
- We need a test suite for the validation tool
- We need a way to identify the elements that trigger errors using a path, line and column number, or both
- We need to be able to suppress errors by error code, path, or a combination of the two using an external file

**Comment by ApiSecRay** [comment]
A start for the test suite is in `specification/tests/invalid`. Run the test suite using `gradlew testValidation`.
